source/tutorials/nix-language.md

```diff
@@ -0,0 +1,1602 @@
+
+```nix
+1 + 2
+```
+
+    3
+
+#### Interactive evaluation
+
+Use [`nix repl`][nix-repl] to evaluate Nix expressions interactively (by typing them on the command line):
+
+```console
+nix repl
+```
+
+    Welcome to Nix 2.5.1. Type :? for help.
+
+```console
+nix-repl> 1 + 2
+```
+
+    3
+
+[nix-repl]: https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-repl.html
+
+#### Evaluating Nix files
+
+Use [`nix-instantiate --eval`][nix-instantiate] to evaluate the expression in a Nix file.
+
+```console
+echo 1 + 2 > file.nix
+
+nix-instantiate --eval file.nix
+```
+
+    3
+
+<details><summary>Detailed explanation</summary>
+
+The first command writes `1 + 2` to a file `file.nix` in the current directory.
+The contents of `file.nix` are now `1 + 2`, which you can check with
+
+```console
+cat file.nix
+```
+
+    1 + 2
+
+The second command runs `nix-instantiate` with the `--eval` option on `file.nix`, which reads the file and evaluates the contained Nix expression.
+The resulting value is printed as output.
+
+`--eval` is required to evaluate the file and do nothing else.
+If `--eval` is omitted, `nix-instantiate` expects the expression in the given file to evaluate to a special data type called *Derivation*, which we will come back to at the end of this guide in [](build-tasks).
+
+If you do not need `file.nix` any more, remove it with
+
+```console
+rm file.nix
+```
+
+</details>
+
+:::{note}
+`nix-instantiate --eval` will evaluate `default.nix` if no file name is specified.
+
+```console
+echo 1 + 2 > default.nix
+
+nix-instantiate --eval
+```
+
+    3
+:::
+
+[nix-instantiate]: https://nixos.org/manual/nix/stable/command-ref/nix-instantiate.html
+
+## Reading the Nix language without fear
+
+You will quickly encounter Nix language expressions that may look very complicated.
+
+As with any programming language, the required amount of Nix language code closely matches the complexity of the problem it is supposed to solve, and reflects how well the problem – and its solution – is understood.
+
+Building software is a complex undertaking, and Nix both exposes and allows managing this complexity with the Nix language.
+
+The purpose of the Nix language is to create *build tasks*: precise descriptions of how contents of existing files are used to derive new files.
+
+:::{important}
+A build task in Nix is called *derivation*.
+:::
+
+The Nix language has only few basic constructs which can be combined arbitrarily:
+
+- primitive data types
+
+  as basic building blocks
+
+- compound data types and functions
+
+  to produce and transform complex data
+
+- name assignment
+
+  to manipulate complex data as units
+
+In addition it allows three side effects:
+
+  - reading files as Nix expressions
+
+    to enable code reuse
+
+  - reading files as *build inputs*
+
+    to capture what build tasks will operate on
+
+  - writing files as *build tasks*
+
+    to keep them for later execution, the *build*
```

Comment:
This is all true, but then again how do we otherwise structure this introduction and still show the relevant mechanisms?

Given the [purpose of the language](https://github.com/notifications?query=backport#discussion_r967533619) it would be pointless to evaluate it without capturing the store derivations. Nix language is not a general-purpose language, there is no point using it outside of that problem domain.

Reply:
(The link you gave doesn't work)

Derivations should be described as the _output_ of a Nix evaluation (it's a somewhat special type of Nix value), not a side effect. So Nix evaluation happens, a value gets produced, which is then used by other parts of Nix to turn them into something `nix-instantiate` (without `--eval`) creates the derivation in the store, and coupling that with `nix-build` also builds it. But the Nix language is not involved past evaluation.